Community/Expire/automation/jsr223/python/community/expire/expire.py

```diff
@@ -103,127 +106,239 @@ def get_config(item_name):
-    is imported from configuration as expire_items_gr.
+    Called when an item configured for expire receives an update.
 
     If the Item updates to an UnDefType, the change is ignored.
     If the Item updates to the same state configured in the expire config any
     expire timer is cancelled.
     If the Item updates to a different state configured in the expire config a
     new timer is created to go off at the configured time in the future.
     """
+    item_name = event.itemName
+
     # Ignore changes to an UnDefType.
     if isinstance(event.itemState, UnDefType):
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
         return
 
-    cfg = get_config(event.itemName)
+    cfg = get_config(item_name, expire.log)
+    if not cfg:
+        expire.log.warn("Skipping expire processing for '{}'"
+                        .format(item_name))
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
+        return
 
     # Cancel the timer when the Item enters the cfg state.
     # Use unicode because there is no degree symbol in ASCII so we can handle
     # Number:Temperature Items.
-    if unicode(items[event.itemName]) == cfg["state"]:
-        if (event.itemName in timers
-                and not timers[event.itemName].hasTerminated()):
-            timers[event.itemName].cancel()
-            del timers[event.itemName]
+    if unicode(items[item_name]) == cfg["state"]:
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
+        return
 
     # Create an expire timer when the Item differs from the end state.
+    expire.log.debug("Setting timer for '{}' with delay {}"
+                     .format(item_name, cfg["time"]))
+    t = (DateTime.now().plusDays(cfg["time"].days)
+                        .plusSeconds(cfg["time"].seconds)
+                        .plusMillis(int(cfg["time"].microseconds/1000)))
+    if (timers[item_name] is not None
+            and not timers[item_name].hasTerminated()):
+        timers[item_name].reschedule(t)
+    else:
+        timers[item_name] = ScriptExecution.createTimer(
+            t,
+            lambda: expired(ir.getItem(item_name),
+                            cfg["type"],
+                            cfg["state"],
+                            expire.log)
+        )
+
+
+def expire_load(event):
+
+    log = logging.getLogger("{}.Expire Load".format(LOG_PREFIX))
+    log.debug("Expire loading...")
+
+    # Keep track of items configured this pass
+    new_items = []
+
+    # Scan for items with valid expire config
+    for item_name in items:
+        cfg = get_config(item_name, log)
+        if cfg:
+            new_items.append(item_name)
+            timers[item_name] = timers.get(item_name, None) # don't clobber
```

Comment:
So you prepopulate the Dict with None so you don't have to use `in` to test for the key. Is there a reason to do it this way? Is it more Pythonic?

Reply:
Not more pythonic, it is used to track removed items.